Adding broadcast session manager methods to process segment

angyangie opened the pull request:
**What does this pull request do? Explain your changes. (required)**

**NOTE**: Still need to test in test-harness

Integrate `BroadcastSessionManager` and associated methods into our existing codebase - this is a rough MVP  implementation of a round-robin strategy for multi-O.

**Specific updates (required)**

- [x] Add `sessManager` to `cxn := &rtmpConnection{}` in `registerConnection`
- [x] Add sessions coming back from `sess := s.startSession(cxn)` in `startSessionListener` into `cxn.sessManager.broadcastSessions = sess`
- [x] Integrate `cxn.sessManager.removeSession(sess)` and `cxn.sessManager.completeSession(sess)` into `processSegment`.
- [x] Add more robust testing around `bsm.checkStatus`, monkey-patching `selectOrchestrator`
- [x] Add to mediaserver_test.go

 **How did you test each of these updates (required)**

- [x] Running nodes locally
- [x] Running nodes in test-harness

**Does this pull request close any open issues?**

More details under the `Task Breakdown: Engineering` section of this document, under `Multi-O breakdown`, step # 4  : 

https://docs.google.com/document/d/1KfEWm0z3svw10TrFH9hiMImkXocl3ba_S-8vmX94HNc/edit#heading=h.1mjk0mji9b8w

**Checklist:**

- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

Commit:
sessions: Add a manager for multiple O sessions.

Commit:
sessions: Return a list from selectOrchestrator.

j0sh commented on server/mediaserver.go:
```diff
@@ -368,7 +369,11 @@ func (s *LivepeerServer) registerConnection(rtmpStrm stream.RTMPVideoStream) (*r
+		sessManager: &BroadcastSessionsManager{
+			broadcastSessions: []*BroadcastSession{},
+			sessLock:          &sync.Mutex{},
+			broadcastSessMap:  make(map[string]*BroadcastSession),
```
I think we can refine how we do this. In particular, let's:

* Clean  up initialization of the BroadcastSessionManager struct
* Eliminate passing around the LivepeerNode and the playlist, which is only there to propagate down to `selectOrchs` which makes it a bit smelly.
* Double down on the concept of abstracting away the `selectOrchs` function for testability / reusability
* Maybe eliminate the need for the code around startSession / startSessionListener.

```golang
type BroadcastSessionsManager  struct {
  sessions []*BroadcastSession{}
  sessionMap map[string]*BroadcastSession
  lock sync.Mutex
  createSessions func() []*BroadcastSession{}
}

func NewSessionManager(n *core.LivepeerNode, pl *core.PlaylistManager) *BroadcastSessionManager {
  m := &BroadcastSessionManager {
    // we really don't have to explicitly define additional fields here; zero values are fine
    // given the field definitions in BroadcastSessionManager above
    sessionMap: make(map[string]*BroadcastSession),
    createSessions: func() []*BroadcastSessionManager {
      return selectOrchestrator(node, pl)
    },
  }
  m.refreshSessions()
  return m
}

...

cxn := &rtmpConnection {
  ...
  sessManager = NewSessionManager(node, pl)
}
```

angyangie replied:
This is very helpful feedback Josh! Always making me a better engineer

j0sh commented on server/broadcast.go:
```diff
@@ -82,7 +88,7 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
 	vProfile := cxn.profile
 
 	cxn.lock.RLock()
```
We won't need to lock here anymore since `selectSession` gives us atomicity

j0sh commented on server/broadcast.go:
```diff
@@ -82,7 +88,7 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
 
 	cxn.lock.RLock()
-	sess := cxn.sess
+	sess := cxn.sessManager.selectSession(node, cxn.pl)
```
This call can probably be moved down to line 117. That allows us to avoid selecting a session until we've completed any uploads to our local OS.

angyangie replied:
True! This prompted me to notice something else - even if we move to line 117, if `uri, err := ios.SaveData(name, seg.Data)` fails (line 134), the session wouldn't be added back to our list (`cxn.sessManager.completeSession(sess)`), to no fault of the O. I'll add `cxn.sessManager.completeSession(sess)` to line 140 right before the return statement.

j0sh commented on server/broadcast.go:
```diff
@@ -150,6 +157,8 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
 				cxn.needOrch <- struct{}{}
 			}
 			return
+		} else {
```
The `else` is not really needed since there is a `return` above. This goes back to the earlier notes about minimizing code under conditionals and maximizing the straight-through code flow.

j0sh commented on server/rpc.go:
```diff
@@ -50,6 +51,88 @@ type Broadcaster interface {
+	}
+}
+
+var selectOrchs = func(n *core.LivepeerNode, pl core.PlaylistManager) ([]*BroadcastSession, error) {
```
I don't think this is used anywhere?

j0sh commented on server/mediaserver.go:
```diff
@@ -59,11 +59,12 @@ var BroadcastJobVideoProfiles = []ffmpeg.VideoProfile{ffmpeg.P240p30fps4x3, ffmp
 var AuthWebhookURL string
 
 type rtmpConnection struct {
```
Do we still need the following fields within `rtmpConnection`? I think we can remove these and the associated code:

* needOrch
* eof
* sess
* lock

We can remove startSession/startSessionListener as well. All this has been subsumed by the session manager.

Note that the startSessionListener has an important property that the session manager doesn't currently have: it only allows for one refresh in-flight at a given time.

angyangie replied:
Really great observations! Helping me be more vigilant about looking at existing code and reviewing which has become superfluous

j0sh commented on server/mediaserver.go:
```diff
@@ -436,7 +439,10 @@ func (s *LivepeerServer) startSessionListener(cxn *rtmpConnection) {
+		cxn.sessManager.broadcastSessions = sess
+		for _, s := range sess {
+			cxn.sessManager.broadcastSessMap[s.OrchestratorInfo.Transcoder] = s
+		}
```
This part has been subsumed by the `refreshSession` call in NewSessionManager so this isn't needed anymore.

j0sh commented on server/rpc.go:
```diff
@@ -50,6 +51,88 @@ type Broadcaster interface {
 	Sign([]byte) ([]byte, error)
 }
 
+type BroadcastSessionsManager struct {
```
We don't currently use this struct within `rpc.go` , should we put it somewhere else? I'm thinking either broadcast.go, mediaserver.go or its own file.

angyangie replied:
hmm don't we use it in `NewSessionManager`, and have associated `BroadcastSessionsManager` methods in the same file - `selectSession`, `removeSession`, etc? Or, do you mean something else?

j0sh replied:
We don't actually call any of this in `rpc.go` though, so I'm not sure that this really needs to live there.

angyangie replied:
Oh I see what you mean. Most of these methods are used in `Broadcast.go` ... perhaps we put them there?

angyangie replied:
`NewSessionManager` could go in mediaserver.go

j0sh commented on server/rpc.go:
```diff
@@ -50,6 +51,88 @@ type Broadcaster interface {
 
+type BroadcastSessionsManager struct {
+	broadcastSessions []*BroadcastSession
+	broadcastSessMap  map[string]*BroadcastSession
```
Naming is a little inconsistent throughout here.

* sessions / sess
* BroadcastSessionManager.broadcastSessions feels a little long and redundant.

FWIW, I prefer brevity in my naming but I'll let you call the scheme as long as it's consistent. Some suggestions: list / table / lock / createSessions

j0sh commented on server/rpc.go:
```diff
@@ -50,6 +51,88 @@ type Broadcaster interface {
+		return
+	}
+
+	// include all O's being used successfully in refreshed list
```
This comment is unclear; I think here it's just exiting early if we receive an empty list from `createSessions` ?

j0sh commented on server/mediaserver_test.go:
```diff
@@ -770,24 +813,25 @@ func TestSessionAssigment_WithTerminatedConnection(t *testing.T) {
 	// sanity: nil session
 	cxn.lock.Lock()
-	assert.Nil(cxn.sess) // nil session
+	assert.Len(cxn.sessManager.broadcastSessions, 0) // nil session
```
Following the sessManager API might be cleaner here:

```golang
sess := cxn.sessManager.selectSession()
assert.Nil(sess)
```

Commit:
Removing startSession and startSessionListener

j0sh commented on server/broadcast_test.go:
```diff
@@ -1,41 +0,0 @@
-package server
```
I'm not sure why this file was removed. Fixed locally; force pushing.

j0sh commented on server/rpc.go:
```diff
@@ -63,19 +68,18 @@ func NewSessionManager(node *core.LivepeerNode, pl core.PlaylistManager) *Broadc
 	numSess := len(bsm.sessList)
 
-	if numSess < minOrchs {
+	if numSess < bsm.minOrchs {
```
This makes testing a bit easier (cleans up some race conditions since refreshSessions will no longer be triggered by default) and also gives us an avenue to handle the case where the session list is naturally shorter than the default minOrchs.

angyangie replied:
True! Makes testing `refreshSessions` much easier given differing #s of registered Orchs

j0sh commented on server/rpc.go:
```diff
@@ -104,11 +108,16 @@ func (bsm *BroadcastSessionsManager) completeSession(sess *BroadcastSession) {
 	}
 }
 
-var selectOrchs = func(n *core.LivepeerNode, pl core.PlaylistManager) ([]*BroadcastSession, error) {
```
This was unused

j0sh commented on server/rpc_test.go:
```diff
@@ -559,14 +559,14 @@ func StubBroadcastSessionsManager() *BroadcastSessionsManager {
 }
 
-func TestSelect(t *testing.T) {
+func TestSelectSession(t *testing.T) {
```
We already have a `TestSelectOrchestrator` which has issues with the match done if running `go test -run Select`. This was mostly problematic because SelectOrchestrator is currently not concurrent safe under `-race`.

j0sh commented on server/rpc_test.go:
```diff
@@ -559,14 +559,14 @@ func StubBroadcastSessionsManager() *BroadcastSessionsManager {
 	assert := assert.New(t)
 	assert.Len(bsm.sessList, 2)
+	expectedSess1 := bsm.sessList[1]
+	expectedSess2 := bsm.sessList[0]
```
Moved this because if the session list happens to be less than length 2 then we'll segfault before the assert.Len

j0sh commented on server/mediaserver_test.go:
```diff
@@ -668,6 +668,11 @@ func TestRegisterConnection(t *testing.T) {
 
 }
 
+// XXX check that broadcast session manager is:
```
Just a reminder for current gaps in test coverage

j0sh commented on server/broadcast.go:
```diff
@@ -22,56 +22,155 @@ import (
+	defer bsm.sessLock.Unlock()
+	numSess := len(bsm.sessList)
+
+	if numSess < bsm.minOrchs {
```
How we handle minOrchs is a bit problematic right now. I think this would cause a refresh to be fired every time an O fails or if the number of Os in the pool is less than `minOrchs` (4 right now?) - this would be common during offchain scenarios and testing.

Also, since 4 is the maximum number of inflight sessions, we want to give ourselves some more overhead there, so let's say 8.

We should fix this with something like `minOrchs=min(OrchestratorPool.Size(), HTTPTimeout / SegLen * 2)`, and set the refresh threshold to something like `refreshThreshold=ceil(minOrchs/2)`. This would allow the refresh threshold to scale along with the pool size up to 8 orchestrators.

O Pool Size | Quantity to Refresh (minOrchs) | Refresh Below (numSessions < refreshThreshold)
-- | -- | --
1 | 1 | 1
2 | 2 | 1
3 | 3 | 2
4 | 4 | 2
5 | 5 | 3
6 | 6 | 3
7 | 7 | 4
8 | 8 | 4
9 | 8 | 4
10 | 8 | 4
11 | 8 | 4
12 | 8 | 4

j0sh committed:
discovery: Add Size() function to O pool.

j0sh commented:
* Rebased onto master and fixed a merge conflict
* Added a few commits addressing https://github.com/livepeer/go-livepeer/pull/806#discussion_r274187906
* Also moved some more code into broadcast.go / broadcast_test.go

j0sh commented on server/broadcast.go:
```diff
@@ -22,63 +23,177 @@ import (
+	sessList []*BroadcastSession
+	sessMap  map[string]*BroadcastSession
+	sessLock *sync.Mutex
+	numOrchs int // how many orchs to request at once
```
Renamed this from `minOrchs` for clarity.

j0sh commented on server/broadcast.go:
```diff
@@ -147,6 +144,7 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment, node *core.Live
 				if monitor.Enabled {
 					monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, monitor.SegmentUploadErrorOS, err.Error())
 				}
+				cxn.sessManager.completeSession(sess)
```
Pushing the O back on the list after an upload failure may be an issue. Consider the following:

```
O1 uses S3
O2 uses direct upload
S3 goes down
Segment 1 comes in
B selects O1
S3 upload fails
B adds O1 back to beginning of list
Segment 2 comes in
B selects O1
S3 upload fails
B adds O1 back to beginning of list
...
```
By removing O1 from the list outright, we can fall back to O2.

angyangie replied:
You're right - `cxn.sessManager.completeSession(sess)` should be removed from here in the event of consistent OS failures

angyangie replied:
Can be replaced with `cxn.sessManager.removeSession(sess)`

j0sh committed:
broadcast: Add rudimentary retries.

j0sh commented on server/broadcast.go:
```diff
@@ -233,18 +232,33 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
-		return
+		return nil
 	}
+	{
```
This is to save the indentation from gofmt in order to make the actual differences in this commit easier to eyeball. We can do a separate commit for cosmetics later.

j0sh commented:
Added rudimentary retries. Missing tests.

j0sh commented:
Note that with retries, we are more prone to receiving segments out of order. Fixing https://github.com/livepeer/go-livepeer/issues/667 is more important now.

angyangie commented:
I'll work on out of order segments now!

Commit:
playlistmanager: Use m3u8.InsertSegment.

The InsertSegment function offers us the following in-built features:

* Thread safety on inserts
* Sliding windows
* Rearranges out of order insertions
* Rejects duplicate sequence numbers
* Sets playlist SeqNo for EXT-X-MEDIA-SEQUENCE

This saves us from having to reimplement some (or all) of these
outside the PlaylistManager. This requires an update to the
ericxtang/m3u8 library to bb992c4f6fa80298ecf993ce21f23eeb14f2070c

See: https://github.com/ericxtang/m3u8/commit/bb992c4f6fa80298ecf993ce21f23eeb14f2070c

j0sh committed:
broadcast: Retry transcoded segment on failed downloads.

Since the download is out of our control, we also remove the
session from the manager lest it fail again on the next retry.

Also recognize that errors saving data to our OS may cause issues
with sig hash checking due to the early exit prior to hashing.
However, such errors do not result in a retry, since they are
most likely local errors, rather than errors from the transcoder.

j0sh commented:
* Rebased on top of latest master.

There were some odd merge conflicts, so fixed those up. This rewrote most of the commit hashes, so delete the local branch prior to fetching the updated remote.

* https://github.com/livepeer/go-livepeer/pull/806/commits/511fdfa24550f0e68280b12be342dd734ab4df86 Thread safe error handling during refreshes.

Also rearranged+commented the fields a bit in the
BroadcastSessionManager struct to hopefully make it more
explicit that basically everything in that struct needs to be
protected by a lock.

* https://github.com/livepeer/go-livepeer/pull/806/commits/7c30eff2dd84b8c0bc70464cc57df89def13e347 Don't select sessions that are no longer in the sessions map.

Retry if the first selected session has been removed from the map.
This may occur if the session is removed while still in the list.
To avoid a runtime search of the session list under lock, simply
fixup the session list at selection time by retrying the selection.

This only becomes an issue in the next commit, where we may be
removing sessions from the map *after* we 'provisionally' add
them back to the list after a successful transcode.

* https://github.com/livepeer/go-livepeer/pull/806/commits/2dcf826485969be9415cc0b93b2237d867ecc751 Retry transcode on failed downloads.

Since the download is out of our control, we also remove the
session from the manager lest it fail again on the next retry.

Also recognize that errors saving data to our OS may cause issues
with sig hash checking due to the early exit prior to hashing.
However, such errors do not result in a retry, since they are
most likely local errors, rather than errors from the transcoder.

j0sh commented:
Rebased onto master.

Added some more logging, fixed up into the "nearest" commits:
* Segment refresh begin / end (debug)
* When a segment begins processing (debug)
* When we fail to select an O (eg, O list is empty) and we drop the segment (normal / info).

Commit:
Integrate sessions into transcode loop.

j0sh commented:
Reworked the commits to make things presentable for merging. The history now reflects more of the finished product with a clearer evolution from start-to-finish.

j0sh reviewed (approved):
OK to :ship: if folks are comfortable with the results from testing this.

angyangie commented:
Just tested again locally, killing O's and T's in rapid succession. No errors in transcoding. I'm comfortable merging this. @ericxtang @darkdarkdragon thoughts?

ericxtang reviewed (approved):
I've tested a bunch yesterday.  I'm good with this.  🚢
